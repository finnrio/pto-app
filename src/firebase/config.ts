import { initializeApp } from 'firebase/app';
import { getAuth } from "firebase/auth";
import { getFirestore } from "firebase/firestore";

// Optionally import the services that you want to use
// import {...} from "firebase/auth";
// import {...} from "firebase/database";
// import {...} from "firebase/firestore";
// import {...} from "firebase/functions";
// import {...} from "firebase/storage";

// Initialize Firebase
const firebaseConfig = {
  apiKey: 'AIzaSyBEjW_646bH_Q4ECQD_g7nXhM_WTDQQZqw',
  authDomain: 'pto-app-30271.firebaseapp.com',
  databaseURL: 'https://pto-app-30271.firebaseio.com',
  projectId: 'pto-app-30271',
  storageBucket: 'pto-app-30271.appspot.com',
  messagingSenderId: '440515616929',
  appId: '1:440515616929:ios:1be65de5cfa7a0048210e0',
  measurementId: '1:440515616929:ios:1be65de5cfa7a0048210e0',
};

export const FIREBASE_APP = initializeApp(firebaseConfig);
// For more information on how to access Firebase in your project,
// see the Firebase documentation: https://firebase.google.com/docs/web/setup#access-firebase

export const FIREBASE_AUTH = getAuth(FIREBASE_APP);
export const FIRESTORE_DB = getFirestore(FIREBASE_APP);