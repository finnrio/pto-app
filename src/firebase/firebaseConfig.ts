import { initializeApp } from "firebase/app";
import { connectAuthEmulator, getAuth } from "firebase/auth";
import { connectFirestoreEmulator, getFirestore } from "firebase/firestore";

// Initialize Firebase
const firebaseConfig = {
  apiKey: "AIzaSyBEjW_646bH_Q4ECQD_g7nXhM_WTDQQZqw",
  authDomain: "pto-app-30271.firebaseapp.com",
  databaseURL: "https://pto-app-30271.firebaseio.com",
  projectId: "pto-app-30271",
  storageBucket: "pto-app-30271.appspot.com",
  messagingSenderId: "440515616929",
  appId: "1:440515616929:ios:1be65de5cfa7a0048210e0",
  measurementId: "1:440515616929:ios:1be65de5cfa7a0048210e0",
};

const FIREBASE_APP = initializeApp(firebaseConfig);

const FIREBASE_AUTH = getAuth(FIREBASE_APP);
const FIRESTORE_DB = getFirestore(FIREBASE_APP);

if (process.env.NODE_ENV === "test") {
  console.log("connecting to emulator");
  connectAuthEmulator(FIREBASE_AUTH, "http://127.0.0.1:9099");
  connectFirestoreEmulator(FIRESTORE_DB, "http://127.0.0.1", 8080);
}

export { FIREBASE_APP, FIREBASE_AUTH, FIRESTORE_DB };
