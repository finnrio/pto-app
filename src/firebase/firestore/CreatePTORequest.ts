import { addDoc, collection, doc, getDoc, updateDoc } from "firebase/firestore";
import { FIREBASE_AUTH, FIRESTORE_DB } from "../firebaseConfig";

function calculateHours(startDate: Date, endDate: Date): number {
  const start = new Date(startDate);
  const end = new Date(endDate);
  if (start > end) {
    throw new Error("Start date must be before end date.");
  }
  let count = 0;
  while (start <= end) {
    const dayOfWeek = start.getDay();
    if (dayOfWeek > 0 && dayOfWeek < 6) {
      count += 1;
    }
    start.setDate(start.getDate() + 1);
  }
  return count * 8; // assumed 8 hours per day
}

export default async function CreatePTORequest(
  startDate: Date,
  endDate: Date,
  reason: string,
  managerId: string,
) {
  const { currentUser } = FIREBASE_AUTH;
  const { NODE_ENV } = process.env;

  // assert not null currentUser
  if (!currentUser) {
    throw new Error("No user logged in");
  }

  // Calculate the number of hours requested
  const requestedHours = calculateHours(startDate, endDate);

  // get the user's data
  const userDoc = doc(FIRESTORE_DB, `${NODE_ENV}`, currentUser.uid);
  const loadedDoc = await getDoc(userDoc);
  if (!loadedDoc.exists()) {
    console.log("No user data found");
    throw new Error("No user data found!");
  }

  // eslint-disable-next-line camelcase
  const { pto_allowance, pto_pending, pto_used } = loadedDoc.data();

  // check if the user has enough PTO hours available
  // eslint-disable-next-line camelcase
  if (pto_pending + pto_used + requestedHours > pto_allowance) {
    console.log("Not enough PTO hours available");
    throw new Error("Not enough PTO hours available");
  }

  const ptoCollection = collection(userDoc, "pto");

  // Add a new document with a generated id.
  await addDoc(ptoCollection, {
    start_date: startDate.toISOString().split("T")[0],
    end_date: endDate.toISOString().split("T")[0],
    reason,
    status: "Pending",
    user_id: currentUser.uid,
    hours: requestedHours,
  });

  // Notify manager of new PTO request
  await addDoc(
    collection(FIRESTORE_DB, `${NODE_ENV}`, managerId, "notifications"),
    { user: currentUser.uid, message: "New PTO request" },
  );

  // Update pto_pending in the user's document and remove from pto_available
  await updateDoc(
    userDoc,
    { pto_pending: pto_pending + requestedHours }, // eslint-disable-line camelcase
  );
}
