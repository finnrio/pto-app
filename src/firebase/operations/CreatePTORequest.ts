import { addDoc, collection, doc, getDoc, setDoc, updateDoc } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";
import { faker } from "@faker-js/faker";

function calculateHours(startDate: Date, endDate: Date): number {
  const start = new Date(startDate);
  const end = new Date(endDate);
  if (start > end) {
    throw new Error("Ending date must be after starting date");
  }
  let count = 0;
  while (start <= end) {
    const dayOfWeek = start.getDay();
    if (dayOfWeek > 0 && dayOfWeek < 6) {
      count += 1;
    }
    start.setDate(start.getDate() + 1);
  }
  return count * 8; // assumed 8 hours per day
}

export default async function CreatePTORequest(
  userId: string,
  startDate: Date,
  endDate: Date,
  reason: string,
) {
  const { NODE_ENV } = process.env;

  // Calculate the number of hours requested
  const requestedHours = calculateHours(startDate, endDate);

  // get the user's data
  const userDoc = doc(FIRESTORE_DB, `${NODE_ENV}`, userId);
  const loadedDoc = await getDoc(userDoc);
  if (!loadedDoc.exists()) {
    console.log("No user data found");
    throw new Error("No user data found");
  }

  // eslint-disable-next-line camelcase
  const { pto_allowance, pto_pending, pto_used } = loadedDoc.data();

  // check if the user has enough PTO hours available
  // eslint-disable-next-line camelcase
  if (pto_pending + pto_used + requestedHours > pto_allowance) {
    console.log("Not enough PTO hours available");
    throw new Error("Not enough PTO hours available");
  }

  const ptoCollection = collection(userDoc, "pto");

  // Add a new document with a generated id.
  const ptoUuid = faker.string.uuid(); // TODO generate a random ID with UUIDv4library
  await setDoc(doc(ptoCollection, ptoUuid), {
    start_date: startDate.toISOString().split("T")[0],
    end_date: endDate.toISOString().split("T")[0],
    reason,
    status: "Pending",
    user_id: userId,
    hours: requestedHours,
  });

  // Update pto_pending in the user's document and remove from pto_available
  await updateDoc(
    userDoc,
    { pto_pending: pto_pending + requestedHours }, // eslint-disable-line camelcase
  );
  return ptoUuid;
}
