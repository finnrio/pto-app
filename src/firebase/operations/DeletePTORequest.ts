import { deleteDoc, doc, getDoc, updateDoc } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function DeletePTORequest(userId: string, ptoId: string) {
  const userDocRef = doc(FIRESTORE_DB, `${process.env.NODE_ENV}`, userId);
  const userData = await getDoc(userDocRef);
  if (!userData.exists()) {
    throw new Error("User not found");
  }

  const ptoDocRef = doc(userDocRef, "pto", ptoId);
  await getDoc(ptoDocRef).then(async (docRef) => {
    if (!docRef.exists()) {
      throw new Error("PTO Request not found");
    } else if (docRef.data().status === "Pending") {
      updateDoc(userDocRef, {
        pto_pending: userData.data().pto_pending - docRef.data().hours,
      });
    } else if (docRef.data().status === "Approved") {
      updateDoc(userDocRef, {
        pto_used: userData.data().pto_used - docRef.data().hours,
      });
    }
    await deleteDoc(
      doc(FIRESTORE_DB, `${process.env.NODE_ENV}`, userId, "pto", ptoId),
    );
  });
}
