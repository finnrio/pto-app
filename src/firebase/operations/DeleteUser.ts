import { deleteDoc, doc } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function DeleteUser(uid: string) {
  await deleteDoc(doc(FIRESTORE_DB, `${process.env.NODE_ENV}`, uid));
}
