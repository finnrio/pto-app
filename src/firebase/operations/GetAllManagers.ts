import { collection, query, where, getDocs } from "firebase/firestore";
import { FIRESTORE_DB } from "../firebaseConfig";

export default async function GetAllManagers(): Promise<any[]> {
  const q = query(
    collection(FIRESTORE_DB, `${process.env.NODE_ENV}`),
    where("role", "==", "Manager"),
  );
  const querySnapshot = await getDocs(q);

  const res: any[] = [];

  querySnapshot.forEach((manager) => {
    res.push({
      value: manager.id,
      label: `${manager.data().first_name} ${manager.data().surname}`,
    });
  });
  return res;
}
