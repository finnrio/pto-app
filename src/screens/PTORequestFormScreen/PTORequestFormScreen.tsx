import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import { Alert, Text, TextInput, TouchableOpacity, View } from "react-native";
import React, { useState } from "react";
import styles from "./styles";
import CreatePTORequest from "../../firebase/firestore/CreatePTORequest";

export default function PTORequestFormScreen({ route, navigation }: any) {
  const [purpose, setPurpose] = useState("");

  function HandleCreatePTORequest() {
    CreatePTORequest(
      new Date(route.params.startDate),
      new Date(route.params.endDate),
      purpose,
      route.params.managerId,
    );
    Alert.alert(
      "PTO request submitted",
      "Your PTO request has been sent to you manager for review",
    );
    navigation.navigate("Calendar");
  }

  const HandleSubmitBtn = () => {
    if (purpose === "") {
      Alert.alert(
        "Purpose is required",
        "Please enter a purpose for your PTO request",
      );
    } else {
      HandleCreatePTORequest();
    }
  };

  const HandleCancelBtn = () => {
    Alert.alert("PTO request cancelled", "Your request has been cancelled");
    navigation.navigate("Calendar");
  };

  return (
    <View style={styles.container}>
      <KeyboardAwareScrollView
        style={{ flex: 1, width: "100%" }}
        keyboardShouldPersistTaps="always"
      >
        <Text style={styles.text}>Start Date</Text>
        <TextInput
          style={styles.input}
          placeholder="Start Date of PTO Request"
          placeholderTextColor="#aaaaaa"
          value={route.params.startDate}
          underlineColorAndroid="transparent"
          editable={false}
          testID="startDate_input"
        />
        <Text style={styles.text}>End Date</Text>
        <TextInput
          style={styles.input}
          placeholder="End Date of PTO Request"
          placeholderTextColor="#aaaaaa"
          value={route.params.endDate}
          underlineColorAndroid="transparent"
          editable={false}
          testID="endDate_input"
        />
        <Text style={styles.text}>Purpose</Text>
        <TextInput
          style={styles.input}
          placeholder="Purpose of PTO Request"
          placeholderTextColor="#aaaaaa"
          onChangeText={(text) => {
            setPurpose(text);
          }}
          value={purpose}
          underlineColorAndroid="transparent"
          autoCapitalize="none"
          testID="purpose_input"
        />
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            HandleSubmitBtn();
          }}
          testID="submit_btn"
        >
          <Text style={styles.buttonTitle}>Submit PTO Request</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            HandleCancelBtn();
          }}
          testID="cancel_btn"
        >
          <Text style={styles.buttonTitle}>Cancel PTO Request</Text>
        </TouchableOpacity>
      </KeyboardAwareScrollView>
    </View>
  );
}
