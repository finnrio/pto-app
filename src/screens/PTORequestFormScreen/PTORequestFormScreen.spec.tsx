import React from "react";
import { Alert } from "react-native";
import {
  cleanup,
  render,
  fireEvent,
  screen,
  act,
} from "@testing-library/react-native";
import PTORequestFormScreen from "./PTORequestFormScreen";
import { UserContext } from "../../context/UserContext";
import { User } from "firebase/auth";
import { createMock } from "@golevelup/ts-jest";

// solution from https://github.com/APSL/react-native-keyboard-aware-scroll-view/issues/493#issuecomment-1023551697
jest.mock("react-native-keyboard-aware-scroll-view", () => ({
  KeyboardAwareScrollView: (props: { children: React.ReactNode }) =>
    props.children,
}));

jest.mock("../../firebase/operations/CreatePTORequest", () => {
  return jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });
});

const mockAuthUser: User = createMock<User>({ uid: "mock_uid" });

const navigation = {
  navigate: jest.fn(),
};

const startDate = "2024-01-01";
const endDate = "2024-01-02";
const route = {
  params: {
    startDate,
    endDate
  }
};

const spyAlert = jest.spyOn(Alert, "alert");

describe("PTORequestFormScreen", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const { toJSON } = render(
      <UserContext.Provider value={mockAuthUser}>
        <PTORequestFormScreen navigation={navigation} route={route}/>
      </UserContext.Provider>
      );
    expect(toJSON()).toMatchSnapshot();
  });
  it("renders the PTO request form with passed in dates", async () => {
    render(
      <UserContext.Provider value={mockAuthUser}>
        <PTORequestFormScreen navigation={navigation} route={route}/>
      </UserContext.Provider>
    );
    expect(screen.getByTestId("startDate_input").props.value).toBe(startDate);
    expect(screen.getByTestId("endDate_input").props.value).toBe(endDate);
    expect(screen.getByTestId("purpose_input").props.value).toBe("");
    expect(screen.getByTestId("submit_btn")).toBeTruthy();
    expect(screen.getByTestId("cancel_btn")).toBeTruthy();
  });
  it("submits the PTO request", async () => {
    const purpose = "mock purpose";
    render(
      <UserContext.Provider value={mockAuthUser}>
        <PTORequestFormScreen navigation={navigation} route={route}/>
      </UserContext.Provider>
    );
    fireEvent.changeText(screen.getByTestId("purpose_input"), purpose);
    fireEvent.press(screen.getByTestId("submit_btn"));
    expect(require("../../firebase/operations/CreatePTORequest")).toHaveBeenCalledWith(
      "mock_uid",
      new Date(startDate),
      new Date(endDate),
      purpose
    );
    expect(navigation.navigate).toHaveBeenCalledWith("Calendar");
  });
  it("alerts the user if CreatePTOEvent fails", async () => {
    const purpose = "mock purpose";
    require("../../firebase/operations/CreatePTORequest").mockImplementationOnce( () => {
      return Promise.reject(new Error("mock error"));
    });
    render(
      <UserContext.Provider value={mockAuthUser}>
        <PTORequestFormScreen navigation={navigation} route={route}/>
      </UserContext.Provider>
    );
    await act(async () => {
      await fireEvent.changeText(screen.getByTestId("purpose_input"), purpose);
      await fireEvent.press(screen.getByTestId("submit_btn"));
    })
    expect(spyAlert).toHaveBeenCalledWith("Error", "mock error");
  });
  it("should alert the user when purpose is empty", async () => {
    render(
      <UserContext.Provider value={mockAuthUser}>
        <PTORequestFormScreen navigation={navigation} route={route}/>
      </UserContext.Provider>
    );
    fireEvent.press(screen.getByTestId("submit_btn"));
    expect(spyAlert).toHaveBeenCalledWith("Purpose is required", "Please enter a purpose for your PTO request");
  });
  it("should return the user to the calendar screen on cancel", async () => {
    render(
      <UserContext.Provider value={mockAuthUser}>
        <PTORequestFormScreen navigation={navigation} route={route}/>
      </UserContext.Provider>
    );
    fireEvent.press(screen.getByTestId("cancel_btn"));
    expect(navigation.navigate).toHaveBeenCalledWith("Calendar");
  });
});