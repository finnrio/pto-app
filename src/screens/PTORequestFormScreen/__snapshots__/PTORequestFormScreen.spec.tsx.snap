// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PTORequestFormScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    Start Date
  </Text>
  <TextInput
    editable={false}
    placeholder="Start Date of PTO Request"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    testID="startDate_input"
    underlineColorAndroid="transparent"
    value="2024-01-01"
  />
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    End Date
  </Text>
  <TextInput
    editable={false}
    placeholder="End Date of PTO Request"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    testID="endDate_input"
    underlineColorAndroid="transparent"
    value="2024-01-02"
  />
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    Purpose
  </Text>
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="Purpose of PTO Request"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    testID="purpose_input"
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#788eec",
        "borderRadius": 5,
        "height": 48,
        "justifyContent": "center",
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 20,
        "opacity": 1,
      }
    }
    testID="submit_btn"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Submit PTO Request
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#788eec",
        "borderRadius": 5,
        "height": 48,
        "justifyContent": "center",
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 20,
        "opacity": 1,
      }
    }
    testID="cancel_btn"
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Cancel PTO Request
    </Text>
  </View>
</View>
`;
